libraries/ESP8266WebServer/src/Parsing-impl.h

```diff
@@ -440,74 +441,61 @@ bool ESP8266WebServerTemplate<ServerType>::_parseForm(ClientType& client, const
-                    _currentUpload->filename.c_str(),
-                    _currentUpload->type.c_str(),
-                    (int)_currentUpload->totalSize);
-                line = client.readStringUntil(0x0D);
-                client.readStringUntil(0x0A);
-                if (line == "--"){
-                  DBGWS("Done Parsing POST\n");
-                  break;
+                if ((strncmp((const char*)boundBuf, "--", 2) == 0) && (strcmp((const char*)(boundBuf + 2), boundary.c_str()) == 0))
+                    break;   //found the boundary, done parsing this file
+                if (first) first = false;   //only add newline characters after the first line
+                else {
+                    _uploadWriteByte('\r');
+                    _uploadWriteByte('\n');
                 }
-                continue;
-              } else {
-                _uploadWriteByte(0x0D);
-                _uploadWriteByte(0x0A);
-                _uploadWriteByte((uint8_t)('-'));
-                _uploadWriteByte((uint8_t)('-'));
-                uint32_t i = 0;
-                while(i < boundary.length()){
-                  _uploadWriteByte(endBuf[i++]);
+                // current line does not contain boundary, upload all bytes in boundary buffer
+                for (int j = 0; j < i; j++)
+                    _uploadWriteByte(boundBuf[j]);
+                // the initial pass (filling up the boundary buffer) did not reach the end of the line. Upload the rest of the line now
+                if (i >= 2 + bLen) {
+                    argByte = _uploadReadByte(client);
+                    while (argByte != '\r') {
+                        if (!client.connected()) return _parseFormUploadAborted();
+                        _uploadWriteByte(argByte);
+                        argByte = _uploadReadByte(client);
+                    }
                 }
-                argByte = _uploadReadByte(client);
-                goto readfile;
-              }
-            } else {
-              _uploadWriteByte(0x0D);
-              goto readfile;
+                _uploadReadByte(client); // '\n'
+            }
+            //Found the boundary string, finish processing this file upload
+            if (_currentHandler && _currentHandler->canUpload(_currentUri))
+                _currentHandler->upload(*this, _currentUri, *_currentUpload);
+            _currentUpload->totalSize += _currentUpload->currentSize;
+            _currentUpload->status = UPLOAD_FILE_END;
+            if (_currentHandler && _currentHandler->canUpload(_currentUri))
+                _currentHandler->upload(*this, _currentUri, *_currentUpload);
+            DBGWS("End File: %s Type: %s Size: %d\n",
+                _currentUpload->filename.c_str(),
+                _currentUpload->type.c_str(),
+                (int)_currentUpload->totalSize);
+            line = client.readStringUntil('\r');
+            client.readStringUntil('\n');
```

Comment:
Is it guaranteed in the specs there will be a `\r\n` terminating the line?  (thus making `\r` as line termination impossible)

Reply:
I couldn't find a specification that mandated line endings but the existing code used '\r\n' so that's what I did. I tested on Windows, Android, and Linux and there were no problems with the line endings.

Reply:
[HTTP/1.0 line break is CRLF](https://tools.ietf.org/html/rfc1945#section-2.2)